### ISSUE TITLE
[go_router] `Hero` animation don't work between pages nested on `ShellRoute`

### ISSUE THREAD

**Issue by 70nybl4nc0** [flutter/flutter#112095]
## Steps to Reproduce

1. Execute `flutter run` on the code sample
2. Tap on the Flutter logo to go to details page.

**Expected results:**

https://user-images.githubusercontent.com/34517439/191557964-89479113-5b15-4463-a49d-75568f99b366.MP4

**Actual results:** 

https://user-images.githubusercontent.com/34517439/191557983-c8cfa1ac-d11c-4a11-a4ed-8b08624ca4dd.MP4

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

final router = GoRouter(
  initialLocation: '/a',
  routes: [
    ShellRoute(
      builder: (context, state, child) => child,
      routes: [
        GoRoute(
          path: '/a',
          builder: (context, state) => const PageA(),
          routes: [
            GoRoute(
              path: 'details',
              builder: (context, state) => const PageADetails(),
            )
          ],
        ),
      ],
    )
  ],
);

void main{
  runApp(
      MaterialApp.router(
        routeInformationParser: router.routeInformationParser,
        routeInformationProvider: router.routeInformationProvider,
        routerDelegate: router.routerDelegate,
      ),
   );
}


class PageA extends StatelessWidget {
  const PageA({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: GestureDetector(
          onTap: () {
            context.go('/a/details');
          },
          child: const Hero(
            tag: 'tag',
            child: SizedBox.square(
              dimension: 100,
              child: FlutterLogo(),
            ),
          ),
        ),
      ),
    );
  }
}

class PageADetails extends StatelessWidget {
  const PageADetails({super.key});

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: Center(
        child: Hero(
          tag: 'tag',
          child: SizedBox.square(
            dimension: 200,
            child: FlutterLogo(),
          ),
        ),
      ),
    );
  }
}

```

</details>

<details>
  <summary>Logs</summary>

```
[No exception or error logs]
```

```
Analyzing app...                                                        
No issues found! (ran in 8.7s)
```

```
[✓] Flutter (Channel stable, 3.3.0, on macOS 12.5 21G72 darwin-x64, locale en-US)
    • Flutter version 3.3.0 on channel stable at /Users/tony/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ffccd96b62 (3 weeks ago), 2022-08-29 17:28:57 -0700
    • Engine revision 5e9e0e0aa8
    • Dart version 2.18.0
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/tony/Library/Android/sdk
    • Platform android-31, build-tools 30.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13F100
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.71.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.46.0

[✓] Connected device (3 available)
    • Tony Xr (mobile) • 00008020-000C11A23C02003A • ios            • iOS 15.4.1 19E258
    • macOS (desktop)  • macos                     • darwin-x64     • macOS 12.5 21G72 darwin-x64
    • Chrome (web)     • chrome                    • web-javascript • Google Chrome 97.0.4692.71

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by jonmountjoy** [comment]
I can confirm as I just discovered this as well.

**Comment by exaby73** [comment]
## Triage Report

### Versions reproducible on

- `Stable (3.3.2)`
- `Master (3.4.0-19.0.pre.442)`

### Info

<details>
<summary>Video</summary>

<img width="300" src="https://user-images.githubusercontent.com/50555895/191748054-0d0be3f4-a488-4b27-ba56-fe25e8895cef.gif">

</details>

<details>
<summary>Code Sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

final router = GoRouter(
  initialLocation: '/a',
  routes: [
    ShellRoute(
      builder: (context, state, child) => child,
      routes: [
        GoRoute(
          path: '/a',
          builder: (context, state) => const PageA(),
          routes: [
            GoRoute(
              path: 'details',
              builder: (context, state) => const PageADetails(),
            )
          ],
        ),
      ],
    )
  ],
);

void main() {
  runApp(
    MaterialApp.router(
      routeInformationParser: router.routeInformationParser,
      routeInformationProvider: router.routeInformationProvider,
      routerDelegate: router.routerDelegate,
    ),
  );
}

class PageA extends StatelessWidget {
  const PageA({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: GestureDetector(
          onTap: () {
            context.go('/a/details');
          },
          child: const Hero(
            tag: 'tag',
            child: SizedBox.square(
              dimension: 100,
              child: FlutterLogo(),
            ),
          ),
        ),
      ),
    );
  }
}

class PageADetails extends StatelessWidget {
  const PageADetails({super.key});

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: Align(
        alignment: Alignment.topLeft,
        child: Hero(
          tag: 'tag',
          child: SizedBox.square(
            dimension: 200,
            child: FlutterLogo(),
          ),
        ),
      ),
    );
  }
}
```

</detail
### Flutter Doctor

<details><summary><code>flutter doctor -v</code> (Stable)</summary>

```
[✓] Flutter (Channel stable, 3.3.2, on Microsoft Windows [Version 10.0.22000.978], locale en-US)
    • Flutter version 3.3.2 on channel stable at C:\Users\Nabeel\fvm\versions\stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e3c29ec00c (8 days ago), 2022-09-14 08:46:55 -0500
    • Engine revision a4ff2c53d8
    • Dart version 2.18.1
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at C:\android-sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2022 17.3.2)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.3.32819.101
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[✓] VS Code (version 1.71.2)
    • VS Code at C:\Users\Nabeel\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.46.0

[✓] Connected device (4 available)
    • sdk gphone64 x86 64 (mobile) • emulator-5554 • android-x64    • Android 13 (API 33) (emulator)
    • Windows (desktop)            • windows       • windows-x64    • Microsoft Windows [Version 10.0.22000.978]
    • Chrome (web)                 • chrome        • web-javascript • unknown
    • Edge (web)                   • edge          • web-javascript • Microsoft Edge 105.0.1343.42

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

<details><summary><code>flutter doctor -v</code> (Master)</summary>

```
[✓] Flutter (Channel master, 3.4.0-19.0.pre.442, on Microsoft Windows [Version 10.0.22000.978], locale en-US)
    • Flutter version 3.4.0-19.0.pre.442 on channel master at C:\Users\Nabeel\fvm\versions\master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 463ee3d58c (83 minutes ago), 2022-09-22 07:08:23 -0400
    • Engine revision ac6cb9f244
    • Dart version 2.19.0 (build 2.19.0-228.0.dev)
    • DevTools version 2.17.0

[✓] Windows Version (Installed version of Windows is version 10 or higher)

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at C:\android-sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files\BraveSoftware\Brave-Browser\Application\brave.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2022 17.3.2)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.3.32819.101
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[✓] VS Code (version 1.71.2)
    • VS Code at C:\Users\Nabeel\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.46.0

[✓] Connected device (4 available)
    • sdk gphone64 x86 64 (mobile) • emulator-5554 • android-x64    • Android 13 (API 33) (emulator)
    • Windows (desktop)            • windows       • windows-x64    • Microsoft Windows [Version 10.0.22000.978]
    • Chrome (web)                 • chrome        • web-javascript • unknown
    • Edge (web)                   • edge          • web-javascript • Microsoft Edge 105.0.1343.42

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by bbeckford** [comment]
Also just discovered this on my first time working with go_router 👍

**Comment by angjelkom** [comment]
I've opened PR https://github.com/flutter/packages/pull/2664

It adds `observers` parameter to the ShellRoute, I made this specifically for cases like this and I can confirm that adding a `HeroController` as an observer to the ShellRoute fixes the Hero animation

**Comment by APDos** [comment]
Just came across this issue as well, is there a temporary workaround while this PR is being resolved? Or is it best to wait?

**Comment by RuckyZucky** [comment]
I found a workaround that works for my case.
You can wrap the page created by the shellroute in a HeroControllerScope widget.
`
ShellRoute(
  pageBuilder: (context, state, child) {
    return MaterialPage(
      child: HeroControllerScope(
        controller: MaterialApp.createMaterialHeroController(),
        child: // your widget here
      ),
    ),
  },
),
`

**Comment by enseitankad0** [comment]
@RuckyZucky Could you provide an example? I can not make it working. Hero animation is fully functional when using Navigator instead of Go Router. Everything works fine when I use Navigator instead of GoRouter's push method when clicking on the blue item.

GoRouter is applied to provide basic routing (Tab 1, Tab2, Tab3)
HeroControllerScope is also in use (acording to the above example)


GOOD (hero animation is visible)
```
 onPressed: () => Navigator.of(context).push<void>(
              MaterialPageRoute(
                builder: (context) => DetailsScreen(
                  label: "Ania2",
                ),
              ),
            ),

```

BAD (hero animation is skipped)
```
onPressed: () => context.go('/a/details', extra: "Ania2"),
```

Here's my router config

```

final goRouter = GoRouter(
  initialLocation: '/login',
  navigatorKey: _rootNavigatorKey,
  routes: [
    GoRoute(
      path: '/login',
      pageBuilder: (context, state) => const NoTransitionPage(child: LoginScreen()),
    ),
    ShellRoute(
      navigatorKey: _shellNavigatorKey,
      builder: (context, state, child) {
        return HeroControllerScope(
          controller: MaterialApp.createMaterialHeroController(),
          child: ScaffoldWithBottomNavigation(
            tabs: tabs,
            child: child,
          ),
        );
      },
      routes: [
        // Settings route config hidden
        GoRoute(
          path: '/a',
          pageBuilder: (context, state) => NoTransitionPage(
            key: state.pageKey,
            child: const HomeScreen(label: 'a', detailsPath: '/a/details'),
          ),
          routes: [
            GoRoute(
              path: 'details',
              builder: (context, state) => DetailsScreen(
                label: state.extra as String,
              ),
            ),
          ],
        ),
        // Second tab routing hidden
      ],
    ),
  ],
);
```



https://user-images.githubusercontent.com/35073987/209879700-a65d49bd-1716-4774-8747-1759ad3ed4e3.mp4

**Comment by RuckyZucky** [comment]
I'm not 100% sure, but I think for a hero to work you need the new page to overlay the old one. So try to use a `context.push` (Or `GoRouter.of(context).push`) instead of the `context.go`

**Comment by raymclee** [comment]
> I'm not 100% sure, but I think for a hero to work you need the new page to overlay the old one. So try to use a `context.push` (Or `GoRouter.of(context).push`) instead of the `context.go`

doesn't work for me with flutter 3.7 and gorouter 6

**Comment by RuckyZucky** [comment]
Very interesting. I tested a little bit and found the reason, why it worked in my case.
I used a `LayoutBuilder` to distinguish small and wide layouts. When I remove it, it stops working.
I don't know why the LayoutBuilder had something to do with this. Maybe I can find something if I dig deeper into the code.
But I updated my example above for everyone getting here in the future and I created a snippet with a whole Flutter App that I tested in Flutter >3.7 and go_router 6.0.1
https://gist.github.com/RuckyZucky/2a5874aeeea02ccf847f82d78fd392a7
